instrumentation-api-semconv/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/http/AlternateUrlSchemeProvider.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.api.instrumenter.http;
+
+import static io.opentelemetry.instrumentation.api.instrumenter.http.ForwardedHeaderParser.extractProtoFromForwardedHeader;
+import static io.opentelemetry.instrumentation.api.instrumenter.http.ForwardedHeaderParser.extractProtoFromForwardedProtoHeader;
+import static io.opentelemetry.instrumentation.api.instrumenter.http.HttpCommonAttributesExtractor.firstHeaderValue;
+
+import io.opentelemetry.instrumentation.api.internal.ConfigPropertiesUtil;
+import java.util.function.Function;
+
+final class AlternateUrlSchemeProvider<REQUEST> implements Function<REQUEST, String> {
+
+  // if set to true, the instrumentation will prefer the scheme from Forwarded/X-Forwarded-Proto
+  // headers over the one extracted from the URL
+  private static final boolean PREFER_FORWARDED_URL_SCHEME =
+      ConfigPropertiesUtil.getBoolean(
+          "otel.instrumentation.http.prefer-forwarded-url-scheme", false);
+
+  private final HttpServerAttributesGetter<REQUEST, ?> getter;
+
+  AlternateUrlSchemeProvider(HttpServerAttributesGetter<REQUEST, ?> getter) {
+    this.getter = getter;
+  }
+
+  @Override
+  public String apply(REQUEST request) {
```

Comment:
New class with no unit test coverage. :( Can you add it? It might also show a use case for non-static helper methods on the parser....

Reply:
All of these classes are actually covered by the existing extractor tests; in fact, the `.http` package has ~94% test coverage. Should I still add separate unit test for all of these extracted things?

Reply:
I mostly think we shouldn't rely on another unit to get our test coverage. Testing via other classes is less robust I think and usually leads to more complicated test code (harder to set up and/or validate).

Reply:
> Testing via other classes is less robust I think and usually leads to more complicated test code (harder to set up and/or validate).

I tend to disagree when it comes to testing extractors: testing the whole thing allows you to easily confirm that it works according to the spec.

I'll add unit tests for the new classes 👍